src/IECoreDelight/Renderer.cpp

```diff
@@ -606,7 +621,8 @@ class DelightAttributes : public IECoreScenePreview::Renderer::AttributesInterfa
 			{
 				if( const ShaderNetwork *shader = reportedCast<const ShaderNetwork>( o, "attribute", name ) )
 				{
-					return shaderCache->get( shader, attributes );
+					ConstShaderNetworkPtr preprocessedNetwork = IECoreDelight::ShaderNetworkAlgo::preprocessedNetwork( shader );
```

Comment:
This is a tricky one, or at least the path to why I made this change was. The problem was when changing the geometry (let's say width of a RectLight) in an interactive render, the light shader was getting dropped. The geometry would change, but the light would go black.

Here's what I believe is happening :
1. First time we get the attributes, we look in the `ShaderCache` for the USD shader.
2. `ShaderCache` is keyed to the USD shader, it's the first time we've seen it, so we create a shader for it.
3. The shader that is created and sent to 3Delight is based on `preprocessedNetwork()`, which also creates the geometry.
4. User changes width.
5. `ShaderCache` looks for the light shader, doesn't find it because the USD shader is new.
6. Like before it does the creation based on `preprocessedNetwork()`. `preprocessedNetwork()` has updated the new geometry. Yay! But the 3Delight shader is the same as before so nothing changes there.
7. BUT! We keyed to the USD shader, so when `ShaderCache::clearUnused()` is called, we find a dangling shader and dutifully delete it.
8. That was our light shader we didn't bother recreating because it was the same as before!
9. The light changes size but turns black.
10. User is sad.

Maybe my mistake was `const string name = "shader:" + preprocessedNetwork->Object::hash().toString();`? Making it `const string name = "shader:" + shaderNetwork->Object::hash().toString();` fixes it. I don't _think_ there are any red flags due to making 3Delight make a new shader in this case?

Reply:
I made this change in https://github.com/GafferHQ/gaffer/pull/5603/commits/a0c61d62e4bb80bef1f2ddfe3839f68389fcb36d. It seems to be working well so far.